### PR TITLE
CI: Add a workflow which automates some labeling

### PR DESCRIPTION
**Pull request by BadMagic100**
## What is this fixing or adding?

First and foremost, a lot of commits. You should probably squash merge this.

There are 2 categories of automated labeling described below.

### Content-based labeling

Several labels are applied based on the content of the PR. When the PR is updated, labels may be added and removed accordingly  with the changes.

* the `is: documentation` label is added to any PRs which only affects things in docs folders, or any README.md file.
* the `affects: webhost` label is added if any change to WebHost.py or any content in the WebHostLib folder.
* the `affects: core` label is added if any non-world file is modified, with a large handful of exceptions. Changes to .py files in worlds/generic are also counted.

### State-based labeling

The `waiting-on:` labels are added and removed during certain PR events.

* When a non-draft PR is opened or reopened, the `waiting-on: peer-review` label is automatically added. The same applies when a draft PR is marked ready for review.
* When a PR is converted to draft or closed, all `waiting-on:` labels are removed.

## How was this tested?
Through numerous PRs against my own fork, the most notable of which is this one: https://github.com/BadMagic100/Archipelago/pull/18

### CODE REVIEW

File: .github/labeler.yml
Comment (Silvris) [inline .github/labeler.yml:13]: CommonClient is considered core (at the moment at least), so we shouldn't block it here. `**/!(*CommonClient).py` would suffice, or putting it in the any-glob-to-any-file.

**Review by Ixrec — APPROVED** [review]
Not sure how much a "peer review" means on a change like this, but this matches what I had in mind after all the labels discussion in #archipelago-dev

File: .github/labeler.yml
Comment (BadMagic100) [reply .github/labeler.yml:13]: Good catch, will add it to the any-to-any list. I was wondering the same about bizhawk client as well actually

**Comment by BadMagic100** [comment]
Realized a couple more things can be done (not sure how I will test it tbh)

* If changes are requested by a review, change to waiting on author and remove all waiting on review tags
* On synchronize, put back into peer review. Possibly only if waiting on author? Though I also suppose if it was a core review which resulted in the change then it should go back into core review and the action can't maintain state like that, so leaving that manual may be best

File: .github/labeler.yml
Comment (Silvris) [reply .github/labeler.yml:13]: BizhawkClient and SNIClient would also fall under this I believe.

Comment (black-sliver) [reply .github/labeler.yml:13]: Bizhawk and SNI are strange. They have their own maintainers, kind of, but the component is shared between multiple worlds. I don't think they are truly core in the same sense as other things (also bizhawk should probably be moved to worlds/_bizhawk and sni possibly to worlds/_snes or something). So I'm fine with them not being labeled as `core`.

File: .github/labeler.yml
Comment (black-sliver) [inline .github/labeler.yml:29]: Can we get newline at end of file?

File: .github/workflows/label-pull-requests.yml
Comment (black-sliver) [inline .github/workflows/label-pull-requests.yml:16]: Do we really need to checkout? I don't see the checked-out source being used.

File: .github/labeler.yml
Comment (BadMagic100) [reply .github/labeler.yml:29]: Sure

File: .github/workflows/label-pull-requests.yml
Comment (BadMagic100) [reply .github/workflows/label-pull-requests.yml:16]: The labeler action docs say to do it, I assume it is used to determine which files are impacted in some way or another since the PR event does not seem to have that info on it

Comment (BadMagic100) [reply .github/workflows/label-pull-requests.yml:16]: Actually looking again it seems I misunderstood the docs, it should be fine without

**Comment by black-sliver** [comment]
Do I understand correctly that the workflow might look like this
1. PR opened -> automatic `waiting-on: peer-review`
2. someone left an "incomplete" approval -> automatic `waiting-on: core-maintainer`
3. core maintainer looks at it and sees it's incomplete -> manual `waiting-on: peer-review`
4. someone else also leaves an approval -> automatic `waiting-on: core maintainer`

?

So realistically people should still peer review if the `waiting-on: peer-review` tag is gone to avoid 3.-4.?

**Comment by BadMagic100** [comment]
There is not currently (and I don't believe there can be)  anything that automatically labels `waiting-on: core-review` assuming >1 peer reviews are desired, since it doesn't seem to be possible to count the number of reviews. I'm also not sure what you mean by "incomplete approval"

**Comment by black-sliver** [comment]
By "incomplete" I mean that someone only tested functionality and not read code, or someone only read the code, but didn't properly test it. Or someone just left an LGTM without further explanation and we don't have enough data to know what that means from that particular reviewer.

**Comment by BadMagic100** [comment]
I think that is solved by keeping it manual when the core review label is applied

**Comment by Ixrec** [comment]
> Though I also suppose if it was a core review which resulted in the change then it should go back into core review and the action can't maintain state like that, so leaving that manual may be best

Yeah, this is why I didn't suggest automatic changes on review. Once there are multiple humans actively interacting with the PR it's very hard for automation to do better, even if github actions could count reviews properly, and especially since there don't seem to be hard rules (yet?) for how many / what kinds of peer review are considered sufficient.

**Comment by BadMagic100** [comment]
Regarding this:
> If changes are requested by a review, change to waiting on author and remove all waiting on review tags

Though I think this would generally be a safe automation, it's not clear how to get what "kind" of review it is from the pull request review event (review.state might be it, but allowed values are not listed so testing it will be near impossible). So I'll leave it that once humans get involved, the changing of the state tags can be manual.

One possible concern is that people without triage role may not be able to change the tags; if that becomes an issue we could investigate having a bot monitor issue_comment events for specific text and change state labels based on that. I think that's something better revisited later though.

**Comment by BadMagic100** [comment]
Addressed comments and tested changes on the same PR BadMagic100#18

**Review by agilbert1412 — APPROVED** [review]

**Review by black-sliver — APPROVED** [review]

**Pull request merged**
